### PR TITLE
cross compile for windows

### PR DESCRIPTION
**Pull request by lanluo-nvidia**
# Description

The cross compile for windows change has added the following new interface:

**Usage 1:**
cross compile and save the exported program to the disk in Linux:
`torch_tensorrt.cross_compile_for_windows(toch.nn.Module, file_path, inputs, …..)`
cross load the exported program from disk in Windows:
`torch_tensorrt.load_cross_compiled_exported_program(file_path)`

**Usage 2:**
export the model to exported program
`ep = torch.export.export(…)`
cross compile the exported program in Linux and return the trt_module:
`trt_module = torch_tensorrt.dynamo.cross_compile_for_windows(ExportedProgram, inputs,…..)`
save the cross compiled trt module to the disk in Linux:
`torch_tensorrt.dynamo.save_cross_compiled_exported_program(trt_module, inputs,…..)`
cross load the exported program from disk in Windows:
`torch_tensorrt.dynamo.load_cross_compiled_exported_program(file_path)`
 
## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: examples/dynamo/cross_runtime_compilation_for_windows.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+"""
+.. _resnet_cross_runtime_compilation_for_windows_example:
+
+cross runtime compilation limitations:
+The cross compile and saved model can only be loaded in Windows, it can no longer be loaded in Linux
+The cross compile and saved model can only be loaded in the same Compute capacity as the Linux which it was cross compiled
```
Comment (HolyWu) [inline examples/dynamo/cross_runtime_compilation_for_windows.py:6]: I think the formal term in CUDA is [compute capability](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capability) rather than `compute capacity`.

**Comment by peri044** [comment]
One idea from Angela to avoid overhead of setting up engines on windows is to handle in `torch_tensorrt.load()`. This modifies and creates a new graph which makes the engines go live.

**Comment by peri044** [comment]
Proposal: 

1) `torch_tensorrt.dynamo.cross_compile` - Input is exported program, output is an exported program.

2) Exporter:
     a) Currently, for x86 compilation, we create a custom op `execute_engine(inputs, TRTEngine)` in the graph and `TRTEngine` gets registered 
          as a `getattr` node in the graph. 
     b) For cross compilation, pack the engine_info into this `getattr` node and introduce `setup_engine` nodes
```py
if cross_compile:
    # Set the packed TRTEngine info as a graph attribute
    setattr(gm, engine_name, trt_module._pack_engine_info())
    packed_engine_node = gm.graph.get_attr(engine_name)
    engine_node = gm.graph.call_function(torch.ops.tensorrt.setup_engine.default,
                                                  (packed_engine_node),)
    
    # The usual workflow ...                                                
    trt_node = gm.graph.call_function(
                torch.ops.tensorrt.execute_engine.default,
                (trt_module_node.args, engine_node),
            )
```

During torch_tensorrt.load(), we setup the engine and remove the setup_engine node

```py
exp_program = torch_tensorrt.load("trt.ep")
gm = exp_program.module()
# Check if the engine is not set up yet
is_engine_live = not any(node.target == torch.ops.tensorrt.setup_engine.default for node in gm.graph.nodes)
if not is_engine_live:
     # Lowering pass which replaces setup_engine nodes into execute_engine nodes
     # The packed engine should be replaced with TRTEngine
     # packed_engine_info = getattr(gm, engine_name)
     # trt_engine = torch.classes.tensorrt.Engine(packed_engine_info)
     # setattr(gm, engine_name, trt_engine)
     # remove setup_engine node
     gm = torch_tensorrt.dynamo._exporter.setup_engine(gm)
     return exp_program
```

This ensures we support loading this via 
1) torch.export.load() - This incurs the engine setup cost for every inference (has setup_engine + execute_engine nodes)
2) torch_tensorrt.load() - This does not have any overhead (only contains execute_engine nodes similar to how we do it on x86 platform)

**Comment by narendasan** [comment]
What are you guys' thoughts on maybe just making the cross_compile load now?

File: examples/dynamo/cross_runtime_compilation_for_windows.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    if platform.system() != "Linux" or platform.architecture()[0] != "64bit":
+        raise ValueError(
+            "cross runtime compiled model for windows can only be compiled in Linux system"
+        )
+    compile_spec = {
+        "debug": True,
+        "min_block_size": 1,
+    }
+    torchtrt.cross_compile_for_windows(
+        model, file_path=args.path, inputs=inputs, **compile_spec
```
Comment (narendasan) [inline examples/dynamo/cross_runtime_compilation_for_windows.py:78]: thoughts on `save_path` vs `file_path`?

File: py/torch_tensorrt/_compile.py [hunk 3]
```diff
@@ -281,6 +293,104 @@ def compile(
...
+    inputs: Optional[Sequence[Input | torch.Tensor | InputTensorSpec]] = None,
+    arg_inputs: Optional[Sequence[Sequence[Any]]] = None,
+    kwarg_inputs: Optional[dict[Any, Any]] = None,
+    enabled_precisions: Optional[Set[torch.dtype | dtype]] = None,
+    **kwargs: Any,
+) -> None:
+    """Compile a PyTorch module using TensorRT in Linux for Inference in Windows
+
+    Takes an existing PyTorch module and a set of settings to configure the compiler
+    and it will convert methods to JIT graphs which call equivalent TensorRT serialized
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:308]: JIT Graphs?

File: py/torch_tensorrt/_compile.py [hunk 4]
```diff
@@ -281,6 +293,104 @@ def compile(
...
+    engine info into the disk in the specified file_path user provided.
+    It will then allow user to load the deserialized model from the disk in Windows.
+    Note: the model cross compiled for windows in Linux environmen can only be loaded
+    in Windows.
+
+    Argument:
+        module (Union(torch.nn.Module,torch.jit.ScriptModule): Source module
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:315]: We accept a script module?

File: py/torch_tensorrt/_compile.py [hunk 5]
```diff
@@ -281,6 +293,104 @@ def compile(
         raise RuntimeError("Module is an unknown format or the ir requested is unknown")
 
 
+def cross_compile_for_windows(
+    module: Any,
+    file_path: str,
+    inputs: Optional[Sequence[Input | torch.Tensor | InputTensorSpec]] = None,
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:299]: We dont support InputTensorSpec for new APIs

File: py/torch_tensorrt/_compile.py [hunk 6]
```diff
@@ -406,6 +516,32 @@ def convert_method_to_trt_engine(
         raise RuntimeError("Module is an unknown format or the ir requested is unknown")
 
 
+def cross_load_in_windows(file_path: str = "") -> Any:
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:519]: Need a docstr here

Comment (narendasan) [reply py/torch_tensorrt/_compile.py:519]: `load_cross_compiled_exported_program`?

Comment (narendasan) [inline py/torch_tensorrt/_compile.py:519]: Should this not be in torch_tensorrt.dynamo and reexported for the top level?

File: py/torch_tensorrt/_utils.py [hunk 7]
```diff
@@ -1,11 +1,37 @@
...
 
 def sanitized_torch_version() -> Any:
     return (
         torch.__version__
         if ".nv" not in torch.__version__
         else torch.__version__.split(".nv")[0]
     )
+
+
+@torch.library.custom_op(
```
Comment (narendasan) [inline py/torch_tensorrt/_utils.py:17]: Lets move this stuff to the runtime submodule

File: py/torch_tensorrt/_utils.py [hunk 8]
```diff
@@ -1,11 +1,37 @@
...
+    name: str,
+    serialized_device_info: str,
+    serialized_engine: str,
+    serialized_in_binding_names: str,
+    serialized_out_binding_names: str,
+    serialized_hardware_compatible: str,
+    serialized_metadata: str,
+    serialized_target_platform: str,
+) -> List[torch.Tensor]:
+    logger.warning(
```
Comment (narendasan) [inline py/torch_tensorrt/_utils.py:32]: Log error if you are just going to error anyway

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 9]
```diff
@@ -47,6 +48,62 @@
...
+        )
+
+    if not file_path:
+        raise ValueError("File path cannot be empty. Please provide a valid file path")
+
+    # enable cross compile for windows
+    kwargs["enable_cross_compile_for_windows"] = True
+
+    # disable the following settings which should not be enabled for cross compile for windows
+    key_sets = {
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:76]: call this unsupported_settings or something

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 10]
```diff
@@ -47,6 +48,62 @@
...
+        "cache_built_engines",
+        "reuse_cached_engines",
+        "custom_engine_cache",
+    }
+    # disable these settings if anything is turned on
+    kwarg_key_sets = set(kwargs.keys())
+    for key in key_sets.intersection(kwarg_key_sets):
+        if kwargs.get(key):
+            logger.warning(
+                f"arg: {key} should not be enabled for cross compile for windows feature, it is ignored."
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:88]: `is not supported for cross-compilation`

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 11]
```diff
@@ -91,6 +148,7 @@ def compile(
     custom_engine_cache: Optional[BaseEngineCache] = _defaults.CUSTOM_ENGINE_CACHE,
     use_explicit_typing: bool = _defaults.USE_EXPLICIT_TYPING,
     use_fp32_acc: bool = _defaults.USE_FP32_ACC,
+    enable_cross_compile_for_windows: bool = _defaults.ENABLE_CROSS_COMPILE_FOR_WINDOWS,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:151]: Does this work? if I use `dynamo.compile` + `enable_cross_compile_for_windows`, will it succeed or hit the issues with setup? We probably want to figure out another way do this this through since this is a user API

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 12]
```diff
@@ -16,6 +17,28 @@
     OutputSpec,
     TensorArgument,
 )
+from torch_tensorrt.dynamo.runtime._TorchTensorRTModule import ENGINE_IDX, NAME_IDX
+
+
+def cross_save_for_windows(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_exporter.py:23]: `save_cross_compiled_win_ep` or something. Make this names consistent across save and load

File: tests/py/dynamo/runtime/test_000_cross_compile_for_windows.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
+import os
```
Comment (narendasan) [inline tests/py/dynamo/runtime/test_000_cross_compile_for_windows.py:1]: Should this be a L0 test? Feel like L2 or L3

File: examples/dynamo/cross_runtime_compilation_for_windows.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
+"""
+.. _resnet_cross_runtime_compilation_for_windows_example:
+
+cross runtime compilation limitations:
+The cross compile and saved model can only be loaded in Windows, it can no longer be loaded in Linux
+The cross compile and saved model can only be loaded in the same Compute capacity as the Linux which it was cross compiled
```
Comment (peri044) [inline examples/dynamo/cross_runtime_compilation_for_windows.py:6]: compute capability*

File: py/torch_tensorrt/_compile.py [hunk 15]
```diff
@@ -281,6 +293,104 @@ def compile(
...
+        arg_inputs = [arg_inputs]  # type: ignore
+
+    # Export the module
+    torchtrt_arg_inputs = prepare_inputs(arg_inputs)
+    torchtrt_kwarg_inputs = prepare_inputs(kwarg_inputs)
+
+    exp_program = dynamo_trace(
+        module, torchtrt_arg_inputs, kwarg_inputs=torchtrt_kwarg_inputs, **kwargs
+    )
+    logger.info("successfully exported the module")
```
Comment (peri044) [inline py/torch_tensorrt/_compile.py:381]: Consider changing this to debug log

File: py/torch_tensorrt/_compile.py [hunk 16]
```diff
@@ -281,6 +293,104 @@ def compile(
...
+
+    # Compile and save the module
+    dynamo_cross_compile_for_windows(
+        exp_program,
+        file_path,
+        arg_inputs=torchtrt_arg_inputs,
+        enabled_precisions=enabled_precisions_set,
+        **kwargs,
+    )
+    logger.info("successfully compiled and saved the module for windows")
```
Comment (peri044) [inline py/torch_tensorrt/_compile.py:391]: Consider changing this to debug log

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 17]
```diff
@@ -16,6 +17,28 @@
...
+    gm: torch.fx.GraphModule,
+    file_path: str,
+) -> None:
+    gm = copy.deepcopy(gm)
+
+    # Inline TensorRT submodules for windows
+    inline_trt_modules_for_windows(gm)
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_exporter.py:30]: I think define `cross_compile` argument in `inline_trt_modules` call and handle the corresponding cross compile logic within it. This way, we don't need to copy the code twice and it's easier to maintain.

Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py:30]: So this save function could effectively be something like 
```py
def save_cross_compiled_exported_program():
        ep = export(gm, cross_compile=True)
        torch.export.save(exp_program, file_path)

def export(gm: torch.fx.GraphModule, cross_compile: bool = False):
        patched_module = transform(gm, cross_compile=cross_compile)
        exp_program = create_trt_exp_program(patched_module)
```

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 18]
```diff
@@ -411,3 +434,119 @@ def inline_trt_modules(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
...
+            getitem_nodes = trt_node.users
+            for idx, getitem_node in enumerate(getitem_nodes):
+                getitem_node.meta["val"] = trt_node.meta["val"][idx]
+
+        # Erase the TRT submodule (call_module) node.
+        gm.graph.erase_node(trt_module_node)
+    return gm
+
+
+def replace_placeholder_node_in_windows(
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_exporter.py:501]: consider naming this to `replace_execute_engine_no_op_node`

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 11 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_compiler.py:151]: Same question - To avoid setting up, this option should disable setting up engines by checking the TorchTensorRTModules don't setup engines during initialization. is this happening ?

Comment (lanluo-nvidia) [reply py/torch_tensorrt/dynamo/_compiler.py]: removed this settings, also added check make sure
dynamo.compile + enable_cross_compile_for_windows will raise the error:
please use the dynamo.cross_compile_for_windows() api.

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 19]
```diff
@@ -411,3 +420,111 @@ def inline_trt_modules(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
...
+        # Insert a call_function node to perform inference on TRT engine
+        with gm.graph.inserting_before(trt_module_node):
+            engine_info = trt_module._pack_engine_info()
+            engine_bytes = engine_info[ENGINE_IDX]
+            engine_info[ENGINE_IDX] = base64.b64encode(engine_bytes).decode("utf-8")
+            # insert the no_placeholder node in the graph which should be replaced to the actual execute_engine node while load in the windows
+            trt_node = gm.graph.call_function(
+                torch.ops.tensorrt.no_op_placeholder_for_execute_engine.default,
+                (trt_module_node.args, *engine_info),
+            )
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_exporter.py:455]: I think this is the only difference b/w inline_trt_modules_for_windows and inline_trt_modules. So merge these two functions and gate them via cross_compile flag

**Review by peri044 — CHANGES_REQUESTED** [review]

**Review by peri044 — APPROVED** [review]
LGTM
